Added support for HuggingFace Hub API to retrieve metadata

mann1x opened the pull request:
Add support for HuggingFace Hub API to retrieve metadata and use it to create the symlinks and create a directory structure compatible with LM Store

tarbard commented:
Thanks so much for your contribution.  I will test it out when I get a chance and merge. 

One question I have from my initial look through - I was wondering why you moved from the 

`manifest_dir = base_dir / 'manifests' / 'registry.ollama.ai'`
approach to
`manifest_dir = Path(f'{base_dir}{separator}manifests{separator}registry.ollama.ai')`
?

tarbard commented:
Also, what does the huggingface addition do?

mann1x commented:
> Thanks so much for your contribution. I will test it out when I get a chance and merge.

I've updated it already, I have to send another PR :)



> One question I have from my initial look through - I was wondering why you moved from the
> 
> `manifest_dir = base_dir / 'manifests' / 'registry.ollama.ai'` approach to `manifest_dir = Path(f'{base_dir}{separator}manifests{separator}registry.ollama.ai')` ?

That's because I added support for Windows; on top of handling the separator I had lots of issues the previous approach.
Many weird quirks to handle with Windows.

> Also, what does the huggingface addition do?

The ultimate goal is to handle all models in ollama.
The problem is when you import your models coming from HuggingFace, like from LMS, into ollama.

They don't have any reference anymore to the HF source.
On top lot of people pushed their preferred model from HF to the ollama library.
The only link left is the sha256 hash of the model.

With the API they can be tracked back (almost) to the original source (or to their target on HF if the original is from ollama).
This way you can ensure you don't download again in LMS a model you imported in ollama.
Tee naming convention as well is kept in the more familiar (for LMS users) HF format.

Next step, I'll introduce updating the ollama models and automatic import in ollama library from LMS.
Later, creation of the LMS preset from ollama metadata, should close the circle.

My only grievance is that the lookup on LMS isn't reliable.
I'm opening a request on HF to extend the API to query models only by their sha256 hash.
Otherwise I have to search the models for the name in ollama and then loop over all the results to search for the ollama sha256 hash.
It's very slow and cumbersome and expensive, hence the caching.
Saw quite some models coming from HF which have been wrongly or differently named and those will never be matched this way.
An absolute waste that can be fixed very easily, let's see.
